runtimes/pythonrt/pythonrt.go

```diff
@@ -423,45 +425,107 @@ func (py *pySvc) setupHealthcheck(ctx context.Context) chan (error) {
-// We split it from Call since Call is also used to execute activities.
-func (py *pySvc) initialCall(ctx context.Context, funcName string, args []sdktypes.Value, kwargs map[string]sdktypes.Value) (sdktypes.Value, error) {
-	if len(args) > 0 {
-		return sdktypes.InvalidValue, errors.New("initial call can't have positional args")
-	}
-
-	defer func() {
-		py.cleanup(ctx)
-		py.log.Info("Python subprocess cleanup after initial call is done")
-	}()
-
-	py.log.Info("initial call", zap.Any("func", funcName))
+func (py *pySvc) eventData(kwargs map[string]sdktypes.Value) ([]byte, error) {
 	event, err := py.kwToEvent(kwargs)
 	if err != nil {
-		py.log.Error("can't convert event", zap.Error(err))
-		return sdktypes.InvalidValue, fmt.Errorf("can't convert: %w", err)
+		return nil, fmt.Errorf("can't convert event: %w", err)
 	}
 
 	keys := slices.Collect(maps.Keys(event))
 	py.log.Info("event", zap.Any("keys", keys))
 
-	eventData, err := json.Marshal(event)
-	if err != nil {
-		return sdktypes.InvalidValue, fmt.Errorf("marshal event: %w", err)
-	}
+	return json.Marshal(event)
+}
 
-	cancellableCtx, cancel := context.WithCancel(ctx)
+// drainPrints drains the print channel at the end of a run.
+func (py *pySvc) drainPrints(ctx context.Context) {
+	ctx, cancel := context.WithTimeout(ctx, time.Second)
 	defer cancel()
+	// flush the rest of the prints and logs.
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case r := <-py.channels.log:
+			py.log.Log(pyLevelToZap(r.level), r.message)
+		case r := <-py.channels.print:
+			if err := py.cbs.Print(ctx, py.runID, r.message); err != nil {
+				py.log.Error("print error", zap.Error(err))
+			}
+			close(r.doneChannel)
+		}
+	}
+}
 
-	if err := py.startRequest(ctx, funcName, eventData); err != nil {
-		return sdktypes.InvalidValue, fmt.Errorf("start request: %w", err)
+func (py *pySvc) tracebackToLocation(traceback []*pbUserCode.Frame) []sdktypes.CallFrame {
+	frames := make([]sdktypes.CallFrame, len(traceback))
+	for i, f := range traceback {
+		var err error
+		frames[i], err = sdktypes.CallFrameFromProto(&sdktypes.CallFramePB{
+			Name: f.Name,
+			Location: &sdktypes.CodeLocationPB{
+				Path: f.Filename,
+				Row:  f.Lineno,
+				Col:  1,
+				Name: f.Name,
+			},
+		})
+		if err != nil {
+			py.log.Warn("can't translate traceback frame", zap.Error(err))
+		}
+	}
+
+	return frames
+}
+
+// Call handles a function call from autokitteh.
+// First used of Call start a workflow, later invocations are activity calls.
+func (py *pySvc) Call(ctx context.Context, v sdktypes.Value, args []sdktypes.Value, kwargs map[string]sdktypes.Value) (sdktypes.Value, error) {
+	fn := v.GetFunction()
+	if !fn.IsValid() {
+		py.log.Error("call - invalid function", zap.Any("function", v))
+		return sdktypes.InvalidValue, fmt.Errorf("%#v is not a function", v)
+	}
+
+	fnName := fn.Name().String()
+	py.log.Info("call", zap.String("func", fnName))
+
+	var runnerHealthChan chan error
+	if py.firstCall {
```

Comment:
one of these days i'd like to get rid of this firstCall thing

Reply:
I'd like to get rid of many things ;)